### PR TITLE
Implement extraction of full frontend from libretro issue #3321

### PR DESCRIPTION
**Pull request by Slick9000**
RetroArch should delete *.rdb files before extrating new ones from Updater.

If there's a faulty/buggy *.rdb in the old database that doesn't exist in the new one, it'll still may create wrong playlist, even with the fixes released.

Nintendo 3DS has a top screen that can run at 800x240.

Enabling RetroArch (and cores) to run in this mode would enable the user to play the game with higher resolution graphics.

Table of data
I just wanted to make some notes about scaling with specific reference to Nintendo 3DS and the X68k core, which sparked this idea:

800×240 top screen (400×240 in 3D mode)
320×200 bottom screen
The interesting thing here is that we can double the horizontal resolution and have some graphics modes be displayable in higher fidelity than otherwise.

Correct X68k aspect ratio means horizontal stretch of 1.375 to get corrected resolution AFAIK.

When using linux and mesa OpenGL core profiles are apparently not supported in RetroArch.

```[INFO] [GL]: Vendor: nouveau, Renderer: NVF1.```
[INFO] [GL]: Version: 3.1 Mesa 18.1.0-devel (git-776e6af879).
Note that mesa recently exposed a 3.1 compatibility context while before they only has 3.0.```

Now nouveau exposes a 4.3 core context by default for supported cards, but this can easily be hacked to be 4.5 which works and the reasons this is not done upstream are not technical.

```$ glxinfo | grep Core```
OpenGL core profile version string: 4.5 (Core Profile) Mesa 18.1.0-devel (git-776e6af879)
However when forcing a 3.1 - 4.5 core context for RetroArch it will not start.

```MESA_GL_VERSION_OVERRIDE=3.1 retroarch````
or

```MESA_GL_VERSION_OVERRIDE=4.5 retroarch```
Retroarch will start if setting a 1.0 - 3.0 compat context.

```MESA_GL_VERSION_OVERRIDE=1.0 retroarch```
or

```MESA_GL_VERSION_OVERRIDE=3.0 retroarch```
I suppose 3.1 should work too now that mesa has exposed it for compatibility contexts which works when not setting the environment variable, but maybe there is an additional mesa bug somewhere...

RetroArch will also work if setting a higher compatibility context.

```MESA_GL_VERSION_OVERRIDE=4.5COMPAT retroarch```
Expected behavior
RetroArch should prefer a core context if available and then fall back to the current compatibility context if its not.

Note that some libretro cores like beetle-psx do use a core context and will work when starting the core directly. For example if the core supports a core context this will work.

```css
MESA_GL_VERSION_OVERRIDE=4.5 retroarch -L /path/to/core.so /path/to/content
Actual behavior
RetroArch doesn't start.

$ LIBGL_DEBUG=1 MESA_DEBUG=1 MESA_GL_VERSION_OVERRIDE=4.5 ./retroarch --verbose
[INFO] RetroArch 1.7.2 (Git 9e724b1412)
[INFO] === Build =======================================
Capabilities: MMX MMXEXT SSE1 SSE2 SSE3 SSSE3 SSE4 SSE4.2 AVX AES 
Built: Apr  1 2018
[INFO] Version: 1.7.2
[INFO] Git: 9e724b1412
[INFO] =================================================
[INFO] Environ SET_PIXEL_FORMAT: RGB565.
[INFO] Redirecting save file to "/media/data/home/games/roms/.saves/retroarch/.srm".
[INFO] Redirecting savestate to "/media/data/home/games/roms/.saves/retroarch/.sstates/.state".
[INFO] Version of libretro API: 1
[INFO] Compiled against API: 1
[INFO] [Audio]: Set audio input rate to: 29970.03 Hz.
[INFO] [Video]: Video @ fullscreen
[INFO] [GLX]: GLX_OML_sync_control and GLX_MESA_swap_control supported, using better swap control method...
[INFO] [GL]: Found GL context: x
[INFO] [GL]: Detecting screen resolution 1680x1050.
[INFO] [XINERAMA]: Xinerama version: 1.1.
[INFO] [XINERAMA]: Xinerama screens: 1.
[INFO] [GLX]: Using Xinerama on screen #0.
[INFO] [GLX]: X = 0, Y = 0, W = 1680, H = 1050.
[INFO] [GLX]: Using windowed fullscreen.
[INFO] [GLX]: Found swap function: glXSwapIntervalMESA.
[INFO] [GLX]: glXSwapInterval(0)
[INFO] [GL]: Vendor: nouveau, Renderer: NVF1.
[INFO] [GL]: Version: 4.5 (Core Profile) Mesa 18.1.0-devel (git-776e6af879).
Mesa: User error: GL_INVALID_ENUM in glEnable(GL_TEXTURE_2D)
Mesa: User error: GL_INVALID_ENUM in glGetString(GL_EXTENSIONS)
[INFO] [GL]: Supported extensions:
[INFO] [GL]: Using resolution 1680x1050
[INFO] [GL]: Default shader backend found: glsl.
[INFO] [Shader driver]: Using GLSL shader backend.
[INFO] [GLSL]: Checking GLSL shader support ...
[WARN] [GL]: Stock GLSL shaders will be used.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GL]: Using 4 textures.
[INFO] [GL]: Loaded 1 program(s).
[ERROR] [GL debug (High, API, Error)]: GL_INVALID_ENUM in glGetString(GL_EXTENSIONS)
[ERROR] [GL debug (High, API, Error)]: GL_INVALID_ENUM in glGetString(GL_EXTENSIONS)
[ERROR] [GL debug (High, API, Error)]: GL_INVALID_ENUM in glGetString(GL_EXTENSIONS)
[ERROR] [GL debug (High, API, Error)]: GL_INVALID_ENUM in glGetString(GL_EXTENSIONS)
[INFO] [udev]: Keyboard #0 (/dev/input/event1).
[INFO] [udev]: Keyboard #1 (/dev/input/event17).
[INFO] [udev]: Keyboard #2 (/dev/input/event18).
[INFO] [udev]: Mouse #0 (/dev/input/event0).
[INFO] [udev]: Mouse #1 (/dev/input/mouse0).
[INFO] [Joypad]: Found joypad driver: "udev".
[INFO] [Font]: Using font rendering backend: freetype.
Mesa: 13 similar GL_INVALID_ENUM errors
Mesa: User error: GL_INVALID_VALUE in glTexImage2D(internalFormat=GL_LUMINANCE_ALPHA)
[ERROR] [GL debug (High, API, Error)]: GL_INVALID_VALUE in glTexImage2D(internalFormat=GL_LUMINANCE_ALPHA)
[ERROR] GL: Invalid enum.
[ERROR] [Video]: Cannot open video driver ... Exiting ...
[ERROR] Fatal error received in: "init_video()"
[ERROR] This core requires a content file, could not load content.```
Note that there are a few bugs here.

First when forcing core context RetroArch fails to find any mesa extensions.

```[ERROR] [GL]: Neither GL_KHR_debug nor GL_ARB_debug_output are implemented. Cannot start GL debugging.```
The debugging can be worked around with this hack, but other issues may exist as a consequence.

```diff --git a/gfx/drivers/gl.c b/gfx/drivers/gl.c
index 720a4b238b..9f5912c793 100644
--- a/gfx/drivers/gl.c
+++ b/gfx/drivers/gl.c
@@ -1667,8 +1667,6 @@ static void DEBUG_CALLBACK_TYPE gl_debug_cb(GLenum source, GLenum type,
 
 static void gl_begin_debug(gl_t *gl)
 {
-   if (gl_check_capability(GL_CAPS_DEBUG))
-   {
 #ifdef HAVE_GL_DEBUG_ES
       glDebugMessageCallbackKHR(gl_debug_cb, gl);
       glDebugMessageControlKHR(GL_DONT_CARE, GL_DONT_CARE, GL_DONT_CARE, 0, NULL, GL_TRUE);
@@ -1678,9 +1676,6 @@ static void gl_begin_debug(gl_t *gl)
       glDebugMessageControl(GL_DONT_CARE, GL_DONT_CARE, GL_DONT_CARE, 0, NULL, GL_TRUE);
       glEnable(GL_DEBUG_OUTPUT_SYNCHRONOUS);
 #endif
-   }
-   else
-      RARCH_ERR("[GL]: Neither GL_KHR_debug nor GL_ARB_debug_output are implemented. Cannot start GL debugging.\n");
 }
 #endif```
 
Second there are these errors reported by RetroArch

[ERROR] [GL debug (High, API, Error)]: GL_INVALID_ENUM in glGetString(GL_EXTENSIONS)
[ERROR] [GL debug (High, API, Error)]: GL_INVALID_ENUM in glGetString(GL_EXTENSIONS)
[ERROR] [GL debug (High, API, Error)]: GL_INVALID_ENUM in glGetString(GL_EXTENSIONS)
[ERROR] [GL debug (High, API, Error)]: GL_INVALID_ENUM in glGetString(GL_EXTENSIONS)```
Here is a backtrace for these issues. - https://pastebin.com/RwaFaa0V
And this hack works around them.

```diff --git a/gfx/common/gl_common.c b/gfx/common/gl_common.c
index 98a4c41c58..049a61a818 100644
--- a/gfx/common/gl_common.c
+++ b/gfx/common/gl_common.c
@@ -61,14 +61,7 @@ void gl_load_texture_image(GLenum target,
       gl_size_format(&internalFormat);
       glTexStorage2DEXT(target, 1, internalFormat, width, height);
    }
-#else
-   if (gl_check_capability(GL_CAPS_TEX_STORAGE) && internalFormat != GL_BGRA_EXT)
-   {
-      gl_size_format(&internalFormat);
-      glTexStorage2D(target, 1, internalFormat, width, height);
-   }
 #endif
-   else
 #endif
    {
 #ifdef HAVE_OPENGLES````
Next.

Mesa: 13 similar GL_INVALID_ENUM errors
Mesa: User error: GL_INVALID_VALUE in glTexImage2D(internalFormat=GL_LUMINANCE_ALPHA)
[ERROR] [GL debug (High, API, Error)]: GL_INVALID_VALUE in glTexImage2D(internalFormat=GL_LUMINANCE_ALPHA)
Here is a backtrace - https://pastebin.com/FNvWpaQV

And from reading these articles and discussing this on irc with @bparker06 it seems the issue is GL_LUMINANCE_ALPHA where core contexts should use GL_RG.

http://retokoradi.com/2014/03/30/opengl-transition-to-core-profile/
http://docs.gl/gl3/glTexImage2D
https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glTexImage2D.xml

This hack also works around those.

diff --git a/gfx/drivers_font/gl_raster_font.c b/gfx/drivers_font/gl_raster_font.c
index c9b43abfaa..3ffc05e823 100644
--- a/gfx/drivers_font/gl_raster_font.c
+++ b/gfx/drivers_font/gl_raster_font.c
@@ -111,8 +111,8 @@ static bool gl_raster_font_upload_atlas_components_4(gl_raster_t *font)
 static bool gl_raster_font_upload_atlas(gl_raster_t *font)
 {
    unsigned i, j;
-   GLint  gl_internal                   = GL_LUMINANCE_ALPHA;
-   GLenum gl_format                     = GL_LUMINANCE_ALPHA;
+   GLint  gl_internal                   = GL_RG;
+   GLenum gl_format                     = GL_RG;
    size_t ncomponents                   = 2;
    uint8_t       *tmp                   = NULL;
 #if defined(GL_VERSION_3_0)
Now what is remaining.

$ LIBGL_DEBUG=1 MESA_DEBUG=1 MESA_GL_VERSION_OVERRIDE=4.5 ./retroarch --verbose
[INFO] RetroArch 1.7.2 (Git ab1e57015f)
[INFO] === Build =======================================
Capabilities: MMX MMXEXT SSE1 SSE2 SSE3 SSSE3 SSE4 SSE4.2 AVX AES 
Built: Apr  1 2018
[INFO] Version: 1.7.2
[INFO] Git: ab1e57015f
[INFO] =================================================
[INFO] Environ SET_PIXEL_FORMAT: RGB565.
[INFO] Redirecting save file to "/media/data/home/games/roms/.saves/retroarch/.srm".
[INFO] Redirecting savestate to "/media/data/home/games/roms/.saves/retroarch/.sstates/.state".
[INFO] Version of libretro API: 1
[INFO] Compiled against API: 1
[INFO] [Audio]: Set audio input rate to: 29970.03 Hz.
[INFO] [Video]: Video @ fullscreen
[INFO] [GLX]: GLX_OML_sync_control and GLX_MESA_swap_control supported, using better swap control method...
[INFO] [GL]: Found GL context: x
[INFO] [GL]: Detecting screen resolution 1680x1050.
[INFO] [XINERAMA]: Xinerama version: 1.1.
[INFO] [XINERAMA]: Xinerama screens: 1.
[INFO] [GLX]: Using Xinerama on screen #0.
[INFO] [GLX]: X = 0, Y = 0, W = 1680, H = 1050.
[INFO] [GLX]: Using windowed fullscreen.
[INFO] [GLX]: Found swap function: glXSwapIntervalMESA.
[INFO] [GLX]: glXSwapInterval(0)
[INFO] [GL]: Vendor: nouveau, Renderer: NVF1.
[INFO] [GL]: Version: 4.5 (Core Profile) Mesa 18.1.0-devel (git-776e6af879).
Mesa: User error: GL_INVALID_ENUM in glEnable(GL_TEXTURE_2D)
Mesa: User error: GL_INVALID_ENUM in glGetString(GL_EXTENSIONS)
[INFO] [GL]: Supported extensions:
[INFO] [GL]: Using resolution 1680x1050
[INFO] [GL]: Default shader backend found: glsl.
[INFO] [Shader driver]: Using GLSL shader backend.
[INFO] [GLSL]: Checking GLSL shader support ...
[WARN] [GL]: Stock GLSL shaders will be used.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GL]: Using 4 textures.
[INFO] [GL]: Loaded 1 program(s).
[INFO] [udev]: Keyboard #0 (/dev/input/event1).
[INFO] [udev]: Keyboard #1 (/dev/input/event17).
[INFO] [udev]: Keyboard #2 (/dev/input/event18).
[INFO] [udev]: Mouse #0 (/dev/input/event0).
[INFO] [udev]: Mouse #1 (/dev/input/mouse0).
[INFO] [Joypad]: Found joypad driver: "udev".
[INFO] [Font]: Using font rendering backend: freetype.
[ERROR] GL: Invalid enum.
[ERROR] [Video]: Cannot open video driver ... Exiting ...
[ERROR] Fatal error received in: "init_video()"
[ERROR] This core requires a content file, could not load content.

### CODE REVIEW

**Comment by LogPod** [comment]
:shipit:

**Comment by anyputer** [comment]
ooo

File: no u [hunk 1]
```diff
@@ -0,0 +1 @@
+ur mom gae
```
Comment (LogPod) [inline no u]: i agree this code style is much better to use

**Comment by Slick9000** [comment]
kek

**Pull request merged**